# Prosper Loans - Data Visualization
## by Ernest Rowe

#### Dataset: [Loan Data from Prosper](https://www.google.com/url?q=https://s3.amazonaws.com/udacity-hosted-downloads/ud651/prosperLoanData.csv&sa=D&ust=1546576456543000)

> This dataset contains 113,937 loans with 81 variables on each loan, including loan amount, borrower rate (or interest rate), current loan status, borrower income, and many others. More information on the variables in this dataset can be found [here](https://www.google.com/url?q=https://docs.google.com/spreadsheet/ccc?key%3D0AllIqIyvWZdadDd5NTlqZ1pBMHlsUjdrOTZHaVBuSlE%26usp%3Dsharing&sa=D&ust=1546576456544000).

In [62]:
# import all packages, set plots to be embedded inline, & set pandas max_columns to none.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
pd.set_option('display.max_columns', None)

> Load in your dataset and describe its properties through the questions below.
Try and motivate your exploration goals through this section.

In [88]:
# Import loan data & display summary of data.
df = pd.read_csv('./data/prosper_loan_data.csv', parse_dates=['ListingCreationDate', 'ClosedDate', 'DateCreditPulled', 
                                                              'FirstRecordedCreditLine', 'LoanOriginationDate'])

display(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
ListingKey                             113937 non-null object
ListingNumber                          113937 non-null int64
ListingCreationDate                    113937 non-null datetime64[ns]
CreditGrade                            28953 non-null object
Term                                   113937 non-null int64
LoanStatus                             113937 non-null object
ClosedDate                             55089 non-null datetime64[ns]
BorrowerAPR                            113912 non-null float64
BorrowerRate                           113937 non-null float64
LenderYield                            113937 non-null float64
EstimatedEffectiveYield                84853 non-null float64
EstimatedLoss                          84853 non-null float64
EstimatedReturn                        84853 non-null float64
ProsperRating (numeric)                84853 non-null float64
Prospe

None

ListingKey  ListingNumber     ListingCreationDate CreditGrade  \
0  1021339766868145413AB3B         193129 2007-08-26 19:09:29.263           C   
1  10273602499503308B223C1        1209647 2014-02-27 08:28:07.900         NaN   
2  0EE9337825851032864889A          81716 2007-01-05 15:00:47.090          HR   
3  0EF5356002482715299901A         658116 2012-10-22 11:02:35.010         NaN   
4  0F023589499656230C5E3E2         909464 2013-09-14 18:38:39.097         NaN   

   Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  LenderYield  \
0    36  Completed 2009-08-14      0.16516        0.1580       0.1380   
1    36    Current        NaT      0.12016        0.0920       0.0820   
2    36  Completed 2009-12-17      0.28269        0.2750       0.2400   
3    36    Current        NaT      0.12528        0.0974       0.0874   
4    36    Current        NaT      0.24614        0.2085       0.1985   

   EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
0                      NaN            NaN              NaN   
1                  0.07960         0.0249          0.05470   
2                      NaN            NaN              NaN   
3                  0.08490         0.0249          0.06000   
4                  0.18316         0.0925          0.09066   

   ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
0                      NaN                   NaN           NaN   
1                      6.0                     A           7.0   
2                      NaN                   NaN           NaN   
3                      6.0                     A           9.0   
4                      3.0                     D           4.0   

   ListingCategory (numeric) BorrowerState     Occupation EmploymentStatus  \
0                          0            CO          Other    Self-employed   
1                          2            CO   Professional         Employed   
2                          0            GA          Other    Not available   
3                         16            GA  Skilled Labor         Employed   
4                          2            MN      Executive         Employed   

   EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
0                       2.0                 True              True   
1                      44.0                False             False   
2                       NaN                False              True   
3                     113.0                 True             False   
4                      44.0                 True             False   

                  GroupKey        DateCreditPulled  CreditScoreRangeLower  \
0                      NaN 2007-08-26 18:41:46.780                  640.0   
1                      NaN 2014-02-27 08:28:14.000                  680.0   
2  783C3371218786870A73D20 2007-01-02 14:09:10.060                  480.0   
3                      NaN 2012-10-22 11:02:32.000                  800.0   
4                      NaN 2013-09-14 18:38:44.000                  680.0   

   CreditScoreRangeUpper FirstRecordedCreditLine  CurrentCreditLines  \
0                  659.0              2001-10-11                 5.0   
1                  699.0              1996-03-18                14.0   
2                  499.0              2002-07-27                 NaN   
3                  819.0              1983-02-28                 5.0   
4                  699.0              2004-02-20                19.0   

   OpenCreditLines  TotalCreditLinespast7years  OpenRevolvingAccounts  \
0              4.0                        12.0                      1   
1             14.0                        29.0                     13   
2              NaN                         3.0                      0   
3              5.0                        29.0                      7   
4             19.0                        49.0                      6   

   OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
0                      

In [89]:
# Duplicate records are identical except for their ProsperScore value.
duplicate_loans = df[df.ListingKey.duplicated()].ListingKey.values
for i in range(5):
    display(df[df.ListingKey == duplicate_loans[i]])

ListingKey  ListingNumber     ListingCreationDate CreditGrade  \
8  0F043596202561788EA13D5        1023355 2013-12-02 10:43:39.117         NaN   
9  0F043596202561788EA13D5        1023355 2013-12-02 10:43:39.117         NaN   

   Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  LenderYield  \
8    36    Current        NaT       0.0762        0.0629       0.0529   
9    36    Current        NaT       0.0762        0.0629       0.0529   

   EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
8                  0.05221         0.0099          0.04231   
9                  0.05221         0.0099          0.04231   

   ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
8                      7.0                    AA           9.0   
9                      7.0                    AA          11.0   

   ListingCategory (numeric) BorrowerState    Occupation EmploymentStatus  \
8                          7            IL  Food Service         Employed   
9                          7            IL  Food Service         Employed   

   EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup GroupKey  \
8                     269.0                 True             False      NaN   
9                     269.0                 True             False      NaN   

     DateCreditPulled  CreditScoreRangeLower  CreditScoreRangeUpper  \
8 2013-12-02 10:43:39                  820.0                  839.0   
9 2013-12-02 10:43:39                  820.0                  839.0   

  FirstRecordedCreditLine  CurrentCreditLines  OpenCreditLines  \
8              1993-04-01                17.0             16.0   
9              1993-04-01                17.0             16.0   

   TotalCreditLinespast7years  OpenRevolvingAccounts  \
8                        32.0                     12   
9                        32.0                     12   

   OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
8                        219.0                   1.0             6.0   
9                        219.0                   1.0             6.0   

   CurrentDelinquencies  AmountDelinquent  DelinquenciesLast7Years  \
8                   0.0               0.0                      0.0   
9                   0.0               0.0                      0.0   

   PublicRecordsLast10Years  PublicRecordsLast12Months  \
8                       0.0                        0.0   
9                       0.0                        0.0   

   RevolvingCreditBalance  BankcardUtilization  AvailableBankcardCredit  \
8                  9906.0                 0.11                  77696.0   
9                  9906.0                 0.11                  77696.0   

   TotalTrades  TradesNeverDelinquent (percentage)  TradesOpenedLast6Months  \
8         29.0                                 1.0                      1.0   
9         29.0                                 1.0                      1.0   

   DebtToIncomeRatio     IncomeRange  IncomeVerifiable  StatedMonthlyIncome  \
8               0.25  $25,000-49,999              True          3333.333333   
9               0.25  $25,000-49,999              True          3333.333333   

                   LoanKey  TotalProsperLoans  TotalProsperPaymentsBilled  \
8  02163700809231365A56A1C                NaN                         NaN   
9  02163700809231365A56A1C                NaN                         NaN   

   OnTimeProsperPayments  ProsperPaymentsLessThanOneMonthLate  \
8                    NaN                                  NaN   
9                    NaN                                  NaN   

   ProsperPaymentsOneMonthPlusLate  ProsperPrincipalBorrowed  \
8                              NaN                       NaN   
9                              NaN                       NaN   

   ProsperPrincipalOutstanding  ScorexChangeAtTimeOfListing  \
8                          NaN                          NaN   
9                          NaN                          NaN   

ListingKey  ListingNumber     ListingCreationDate  \
998    0A0635972629771021E38F3        1055932 2013-12-18 14:25:17.053   
999    0A0635972629771021E38F3        1055932 2013-12-18 14:25:17.053   
57494  0A0635972629771021E38F3        1055932 2013-12-18 14:25:17.053   

      CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
998           NaN    60    Current        NaT       0.1771        0.1535   
999           NaN    60    Current        NaT       0.1771        0.1535   
57494         NaN    60    Current        NaT       0.1771        0.1535   

       LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
998         0.1435                  0.13507         0.0599          0.07517   
999         0.1435                  0.13507         0.0599          0.07517   
57494       0.1435                  0.13507         0.0599          0.07517   

       ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
998                        5.0                     B           6.0   
999                        5.0                     B           7.0   
57494                      5.0                     B           8.0   

       ListingCategory (numeric) BorrowerState Occupation EmploymentStatus  \
998                            1            MD      Other         Employed   
999                            1            MD      Other         Employed   
57494                          1            MD      Other         Employed   

       EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
998                       100.0                 True             False   
999                       100.0                 True             False   
57494                     100.0                 True             False   

      GroupKey    DateCreditPulled  CreditScoreRangeLower  \
998        NaN 2014-02-08 11:39:13                  760.0   
999        NaN 2014-02-08 11:39:13                  760.0   
57494      NaN 2014-02-08 11:39:13                  760.0   

       CreditScoreRangeUpper FirstRecordedCreditLine  CurrentCreditLines  \
998                    779.0              1994-11-01                16.0   
999                    779.0              1994-11-01                16.0   
57494                  779.0              1994-11-01                16.0   

       OpenCreditLines  TotalCreditLinespast7years  OpenRevolvingAccounts  \
998               16.0                        34.0                     15   
999               16.0                        34.0                     15   
57494             16.0                        34.0                     15   

       OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
998                          623.0                   1.0             5.0   
999                          623.0                   1.0             5.0   
57494                        623.0                   1.0             5.0   

       CurrentDelinquencies  AmountDelinquent  DelinquenciesLast7Years  \
998                     0.0               0.0                      0.0   
999                     0.0               0.0                      0.0   
57494                   0.0               0.0                      0.0   

       PublicRecordsLast10Years  PublicRecordsLast12Months  \
998                         0.0                        0.0   
999                         0.0                        0.0   
57494                       0.0                        0.0   

       RevolvingCreditBalance  BankcardUtilization  AvailableBankcardCredit  \
998                    8028.0                 0.27                  14190.0   
999                    8028.0                 0.27                  14190.0   
57494                  8028.0                 0.27                  14190.0   

       TotalTrades  TradesNeverDelinquent (percentage)  \
998           32.0                                0.96   
999           32.0                                0.96   
57494        

ListingKey  ListingNumber     ListingCreationDate  \
1554  2D2635984503681189056B4        1097584 2014-01-11 07:50:08.037   
2539  2D2635984503681189056B4        1097584 2014-01-11 07:50:08.037   

     CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
1554         NaN    60    Current        NaT      0.14528        0.1224   
2539         NaN    60    Current        NaT      0.14528        0.1224   

      LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
1554       0.1124                  0.10731         0.0424          0.06491   
2539       0.1124                  0.10731         0.0424          0.06491   

      ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
1554                      5.0                     B           9.0   
2539                      5.0                     B          10.0   

      ListingCategory (numeric) BorrowerState  \
1554                          1            NJ   
2539                          1            NJ   

                             Occupation EmploymentStatus  \
1554  Police Officer/Correction Officer         Employed   
2539  Police Officer/Correction Officer         Employed   

      EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
1554                     107.0                 True             False   
2539                     107.0                 True             False   

     GroupKey    DateCreditPulled  CreditScoreRangeLower  \
1554      NaN 2014-03-07 10:04:00                  720.0   
2539      NaN 2014-03-07 10:04:00                  720.0   

      CreditScoreRangeUpper FirstRecordedCreditLine  CurrentCreditLines  \
1554                  739.0              1996-07-01                19.0   
2539                  739.0              1996-07-01                19.0   

      OpenCreditLines  TotalCreditLinespast7years  OpenRevolvingAccounts  \
1554             18.0                        27.0                     15   
2539             18.0                        27.0                     15   

      OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
1554                       1024.0                   1.0             5.0   
2539                       1024.0                   1.0             5.0   

      CurrentDelinquencies  AmountDelinquent  DelinquenciesLast7Years  \
1554                   0.0               0.0                      0.0   
2539                   0.0               0.0                      0.0   

      PublicRecordsLast10Years  PublicRecordsLast12Months  \
1554                       0.0                        0.0   
2539                       0.0                        0.0   

      RevolvingCreditBalance  BankcardUtilization  AvailableBankcardCredit  \
1554                 20802.0                  0.4                  26503.0   
2539                 20802.0                  0.4                  26503.0   

      TotalTrades  TradesNeverDelinquent (percentage)  \
1554         26.0                                 1.0   
2539         26.0                                 1.0   

      TradesOpenedLast6Months  DebtToIncomeRatio IncomeRange  \
1554                      1.0               0.18   $100,000+   
2539                      1.0               0.18   $100,000+   

      IncomeVerifiable  StatedMonthlyIncome                  LoanKey  \
1554              True         10833.333333  DC303708663304668E74587   
2539              True         10833.333333  DC303708663304668E74587   

      TotalProsperLoans  TotalProsperPaymentsBilled  OnTimeProsperPayments  \
1554                NaN                         NaN                    NaN   
2539                NaN                         NaN                    NaN   

      ProsperPaymentsLessThanOneMonthLate  ProsperPaymentsOneMonthPlusLate  \
1554                                  NaN                              NaN   
2539                                  NaN                              NaN   

      ProsperPrincipalBorrowed 

ListingKey  ListingNumber     ListingCreationDate  \
1803  4B7E3590031274113F1FD34         927246 2013-09-26 09:35:36.317   
4942  4B7E3590031274113F1FD34         927246 2013-09-26 09:35:36.317   

     CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
1803         NaN    36    Current        NaT      0.35356        0.3134   
4942         NaN    36    Current        NaT      0.35356        0.3134   

      LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
1803       0.3034                  0.27598         0.1625          0.11348   
4942       0.3034                  0.27598         0.1625          0.11348   

      ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
1803                      1.0                    HR           3.0   
4942                      1.0                    HR           2.0   

      ListingCategory (numeric) BorrowerState Occupation EmploymentStatus  \
1803                          7            MI    Laborer         Employed   
4942                          7            MI    Laborer         Employed   

      EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
1803                     336.0                 True             False   
4942                     336.0                 True             False   

     GroupKey    DateCreditPulled  CreditScoreRangeLower  \
1803      NaN 2013-09-26 09:35:36                  640.0   
4942      NaN 2013-09-26 09:35:36                  640.0   

      CreditScoreRangeUpper FirstRecordedCreditLine  CurrentCreditLines  \
1803                  659.0              1978-09-01                 7.0   
4942                  659.0              1978-09-01                 7.0   

      OpenCreditLines  TotalCreditLinespast7years  OpenRevolvingAccounts  \
1803              6.0                        39.0                      6   
4942              6.0                        39.0                      6   

      OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
1803                        129.0                   3.0             9.0   
4942                        129.0                   3.0             9.0   

      CurrentDelinquencies  AmountDelinquent  DelinquenciesLast7Years  \
1803                   0.0               0.0                     56.0   
4942                   0.0               0.0                     56.0   

      PublicRecordsLast10Years  PublicRecordsLast12Months  \
1803                       1.0                        0.0   
4942                       1.0                        0.0   

      RevolvingCreditBalance  BankcardUtilization  AvailableBankcardCredit  \
1803                   943.0                 0.51                    798.0   
4942                   943.0                 0.51                    798.0   

      TotalTrades  TradesNeverDelinquent (percentage)  \
1803         31.0                                0.56   
4942         31.0                                0.56   

      TradesOpenedLast6Months  DebtToIncomeRatio     IncomeRange  \
1803                      2.0               0.07  $50,000-74,999   
4942                      2.0               0.07  $50,000-74,999   

      IncomeVerifiable  StatedMonthlyIncome                  LoanKey  \
1803              True          4583.333333  8D84369399064302947D8B5   
4942              True          4583.333333  8D84369399064302947D8B5   

      TotalProsperLoans  TotalProsperPaymentsBilled  OnTimeProsperPayments  \
1803                NaN                         NaN                    NaN   
4942                NaN                         NaN                    NaN   

      ProsperPaymentsLessThanOneMonthLate  ProsperPaymentsOneMonthPlusLate  \
1803                                  NaN                              NaN   
4942                                  NaN                              NaN   

      ProsperPrincipalBorrowed  ProsperPrincipalOutstanding  \
1803                       NaN                       

ListingKey  ListingNumber     ListingCreationDate  \
5811  94B035909835592957503E6         920397 2013-09-24 13:27:48.287   
5812  94B035909835592957503E6         920397 2013-09-24 13:27:48.287   

     CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
5811         NaN    36    Current        NaT      0.24205        0.2045   
5812         NaN    36    Current        NaT      0.24205        0.2045   

      LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
5811       0.1945                   0.1797         0.0899           0.0898   
5812       0.1945                   0.1797         0.0899           0.0898   

      ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
5811                      4.0                     C           3.0   
5812                      4.0                     C           4.0   

      ListingCategory (numeric) BorrowerState Occupation EmploymentStatus  \
5811                          1            GA        NaN            Other   
5812                          1            GA        NaN            Other   

      EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
5811                       0.0                 True             False   
5812                       0.0                 True             False   

     GroupKey    DateCreditPulled  CreditScoreRangeLower  \
5811      NaN 2013-10-30 17:36:21                  700.0   
5812      NaN 2013-10-30 17:36:21                  700.0   

      CreditScoreRangeUpper FirstRecordedCreditLine  CurrentCreditLines  \
5811                  719.0              1986-02-10                19.0   
5812                  719.0              1986-02-10                19.0   

      OpenCreditLines  TotalCreditLinespast7years  OpenRevolvingAccounts  \
5811             18.0                        35.0                     16   
5812             18.0                        35.0                     16   

      OpenRevolvingMonthlyPayment  InquiriesLast6Months  TotalInquiries  \
5811                       1383.0                   0.0             0.0   
5812                       1383.0                   0.0             0.0   

      CurrentDelinquencies  AmountDelinquent  DelinquenciesLast7Years  \
5811                   0.0               0.0                      0.0   
5812                   0.0               0.0                      0.0   

      PublicRecordsLast10Years  PublicRecordsLast12Months  \
5811                       0.0                        0.0   
5812                       0.0                        0.0   

      RevolvingCreditBalance  BankcardUtilization  AvailableBankcardCredit  \
5811                 47018.0                 0.96                   1513.0   
5812                 47018.0                 0.96                   1513.0   

      TotalTrades  TradesNeverDelinquent (percentage)  \
5811         34.0                                 1.0   
5812         34.0                                 1.0   

      TradesOpenedLast6Months  DebtToIncomeRatio     IncomeRange  \
5811                      2.0               0.62  $50,000-74,999   
5812                      2.0               0.62  $50,000-74,999   

      IncomeVerifiable  StatedMonthlyIncome                  LoanKey  \
5811              True          4166.666667  129C369881799628360C904   
5812              True          4166.666667  129C369881799628360C904   

      TotalProsperLoans  TotalProsperPaymentsBilled  OnTimeProsperPayments  \
5811                NaN                         NaN                    NaN   
5812                NaN                         NaN                    NaN   

      ProsperPaymentsLessThanOneMonthLate  ProsperPaymentsOneMonthPlusLate  \
5811                                  NaN                              NaN   
5812                                  NaN                              NaN   

      ProsperPrincipalBorrowed  ProsperPrincipalOutstanding  \
5811                       NaN                       

In [90]:
# Remove duplicates & drop PrsoperScore column. Remaining entries should have unique ListingKeys.
df = df[~df.ListingKey.duplicated()].drop(columns='ProsperScore').reset_index(drop=True)
df.head(10)

ListingKey  ListingNumber     ListingCreationDate CreditGrade  \
0  1021339766868145413AB3B         193129 2007-08-26 19:09:29.263           C   
1  10273602499503308B223C1        1209647 2014-02-27 08:28:07.900         NaN   
2  0EE9337825851032864889A          81716 2007-01-05 15:00:47.090          HR   
3  0EF5356002482715299901A         658116 2012-10-22 11:02:35.010         NaN   
4  0F023589499656230C5E3E2         909464 2013-09-14 18:38:39.097         NaN   
5  0F05359734824199381F61D        1074836 2013-12-14 08:26:37.093         NaN   
6  0F0A3576754255009D63151         750899 2013-04-12 09:52:56.147         NaN   
7  0F1035772717087366F9EA7         768193 2013-05-05 06:49:27.493         NaN   
8  0F043596202561788EA13D5        1023355 2013-12-02 10:43:39.117         NaN   
9  0F123545674891886D9F106         587746 2012-05-10 07:04:01.577         NaN   

   Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  LenderYield  \
0    36  Completed 2009-08-14      0.16516        0.1580       0.1380   
1    36    Current        NaT      0.12016        0.0920       0.0820   
2    36  Completed 2009-12-17      0.28269        0.2750       0.2400   
3    36    Current        NaT      0.12528        0.0974       0.0874   
4    36    Current        NaT      0.24614        0.2085       0.1985   
5    60    Current        NaT      0.15425        0.1314       0.1214   
6    36    Current        NaT      0.31032        0.2712       0.2612   
7    36    Current        NaT      0.23939        0.2019       0.1919   
8    36    Current        NaT      0.07620        0.0629       0.0529   
9    60    Current        NaT      0.27462        0.2489       0.2389   

   EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
0                      NaN            NaN              NaN   
1                  0.07960         0.0249          0.05470   
2                      NaN            NaN              NaN   
3                  0.08490         0.0249          0.06000   
4                  0.18316         0.0925          0.09066   
5                  0.11567         0.0449          0.07077   
6                  0.23820         0.1275          0.11070   
7                  0.17830         0.0799          0.09840   
8                  0.05221         0.0099          0.04231   
9                  0.23320         0.0890          0.14420   

   ProsperRating (numeric) ProsperRating (Alpha)  ListingCategory (numeric)  \
0                      NaN                   NaN                          0   
1                      6.0                     A                          2   
2                      NaN                   NaN                          0   
3                      6.0                     A                         16   
4                      3.0                     D                          2   
5                      5.0                     B                          1   
6                      2.0                     E                          1   
7                      4.0                     C                          2   
8                      7.0                    AA                          7   
9                      4.0                     C                          1   

  BorrowerState      Occupation EmploymentStatus  EmploymentStatusDuration  \
0            CO           Other    Self-employed                       2.0   
1            CO    Professional         Employed                      44.0   
2            GA           Other    Not available                       NaN   
3            GA   Skilled Labor         Employed                     113.0   
4            MN       Executive         Employed                      44.0   
5            NM    Professional         Employed                      82.0   
6            KS  Sales - Retail         Employed                     172.0   
7            CA         Laborer         Employed                     103.0   
8            IL    Food Service         Employed                 

In [91]:
# CreditGrade was replaced by ProsperRating (Alpha) for loans originated after July 2009. Create a combined column with all
    # ratings.
df['ProsperCreditGrade'] = df['ProsperRating (Alpha)'].astype(str).str.replace('nan', '') +\
                           df['CreditGrade'].astype(str).str.replace('nan', '')

df['ProsperCreditGrade'].head()

0     C
1     A
2    HR
3     A
4     D
Name: ProsperCreditGrade, dtype: object

In [92]:
# Fill blank spaces with N/A.
df['ProsperCreditGrade'] = df.ProsperCreditGrade.replace('', 'N/A')
df['ProsperCreditGrade'].unique()

array(['C', 'A', 'HR', 'D', 'B', 'E', 'AA', 'N/A', 'NC'], dtype=object)

In [93]:
# Convert Term column from months to years.
df['Term'] = df.Term.astype(str).str.replace('12', '1 Year').str.replace('36', '3 Years').str.replace('60', '5 Years')
df['Term'].unique()

array(['3 Years', '5 Years', '1 Year'], dtype=object)

In [97]:
# Convert ProsperCreditGrade & LoanStatus to categorical dtype.
ordinal_var_dict = {'ProsperCreditGrade': ['NC', 'HR', 'E', 'D', 'C', 'B', 'A', 'AA', 'N/A'],
                    'LoanStatus': ['Canceled', 'Chargedoff', 'Defaulted', 'Past Due (>120 days)', 'Past Due (91-120 days)',
                                   'Past Due (61-90 days)', 'Past Due (31-60 days)', 'Past Due (16-30 days)',
                                   'Past Due (1-15 days)', 'Current', 'FinalPaymentInProgress', 'Completed'],
                    'Term': ['1 Year', '3 Years', '5 Years']}

for var in ordinal_var_dict:
    ordered_var = pd.api.types.CategoricalDtype(ordered=True, categories=ordinal_var_dict[var])
    df[var] = df[var].astype(ordered_var)

In [98]:
# Confirm dtype changed.
df.dtypes

ListingKey                                     object
ListingNumber                                   int64
ListingCreationDate                    datetime64[ns]
CreditGrade                                    object
Term                                         category
LoanStatus                                   category
ClosedDate                             datetime64[ns]
BorrowerAPR                                   float64
BorrowerRate                                  float64
LenderYield                                   float64
EstimatedEffectiveYield                       float64
EstimatedLoss                                 float64
EstimatedReturn                               float64
ProsperRating (numeric)                       float64
ProsperRating (Alpha)                          object
ListingCategory (numeric)                       int64
BorrowerState                                  object
Occupation                                     object
EmploymentStatus            

In [96]:
# Confirm changes.
display(df.info())
df.nunique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113066 entries, 0 to 113065
Data columns (total 81 columns):
ListingKey                             113066 non-null object
ListingNumber                          113066 non-null int64
ListingCreationDate                    113066 non-null datetime64[ns]
CreditGrade                            28953 non-null object
Term                                   113066 non-null object
LoanStatus                             113061 non-null category
ClosedDate                             55076 non-null datetime64[ns]
BorrowerAPR                            113041 non-null float64
BorrowerRate                           113066 non-null float64
LenderYield                            113066 non-null float64
EstimatedEffectiveYield                83982 non-null float64
EstimatedLoss                          83982 non-null float64
EstimatedReturn                        83982 non-null float64
ProsperRating (numeric)                83982 non-null float64
Pro

None

ListingKey                             113066
ListingNumber                          113066
ListingCreationDate                    113064
CreditGrade                                 8
Term                                        3
LoanStatus                                 11
ClosedDate                               2802
BorrowerAPR                              6677
BorrowerRate                             2294
LenderYield                              2283
EstimatedEffectiveYield                  1582
EstimatedLoss                             120
EstimatedReturn                          1476
ProsperRating (numeric)                     7
ProsperRating (Alpha)                       7
ListingCategory (numeric)                  21
BorrowerState                              51
Occupation                                 67
EmploymentStatus                            8
EmploymentStatusDuration                  605
IsBorrowerHomeowner                         2
CurrentlyInGroup                  

In [87]:
# Descriptive statistics of numeric variables.
df.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.130660e+05  113066.000000  113041.000000  113066.000000   
mean    6.248036e+05       3.400014       0.218980       0.192946   
std     3.273262e+05       0.868460       0.080483       0.074917   
min     4.000000e+00       1.000000       0.006530       0.000000   
25%     3.976670e+05       3.000000       0.156290       0.134000   
50%     5.978385e+05       3.000000       0.209840       0.184000   
75%     8.849650e+05       3.000000       0.283860       0.250600   
max     1.255725e+06       5.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113066.000000             83982.000000   83982.000000     83982.000000   
mean        0.182882                 0.168870       0.080417         0.096246   
std         0.074612                 0.068607       0.046851         0.030475   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124500                 0.116000       0.042400         0.074630   
50%         0.174000                 0.161620       0.072400         0.092110   
75%         0.240600                 0.225300       0.112000         0.117100   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ListingCategory (numeric)  \
count             83982.000000              113066.000000   
mean                  4.068705                   2.776838   
std                   1.675442                   3.998188   
min                   1.000000                   0.000000   
25%                   3.000000                   1.000000   
50%                   4.000000                   1.000000   
75%                   5.000000                   3.000000   
max                   7.000000                  20.000000   

       EmploymentStatusDuration  CreditScoreRangeLower  CreditScoreRangeUpper  \
count             105441.000000          112475.000000          112475.000000   
mean                  96.060584             685.524961             704.524961   
std                   94.432241              66.635895              66.635895   
min                    0.000000               0.000000              19.000000   
25%                   26.000000             660.000000             679.000000   
50%                   67.000000             680.000000             699.000000   
75%                  137.000000             720.000000             739.000000   
max                  755.000000             880.000000             899.000000   

       CurrentCreditLines  OpenCreditLines  TotalCreditLinespast7years  \
count       105462.000000    105462.000000               112369.000000   
mean            10.307058         9.248336                   26.746380   
std              5.458196         5.021249                   13.643731   
min              0.000000         0.000000                    2.000000   
25%              7.000000         6.000000                   17.000000   
50%             10.000000         8.000000                   25.000000   
75%             13.000000        12.000000                   35.000000   
max             59.000000        54.000000                  136.000000   

       OpenRevolvingAccounts  OpenRevolvingMonthlyPayment  \
count          113066.000000                113066.000000   
mean                6.959466                   397.820998   
std                 4.629303                   447.657266   
min                 0.000000                     0.000000   
25%                 4.000000                   113.000000   
50%                 6.000000                   271.000000   
75%                 9.000000                   525.000000   
max                51.000000                 14985.000000   

       InquiriesLast6Months  TotalInquiries  CurrentDelinquencies  \
count         112369.000000   111907.000000         112369.000000   
mean               1

### What is the structure of your dataset?

> There are 113066 loans in the cleaned dataset with 81 features in total. This far too may features to explore, so I'll only be focused on the most important ones. Most variables in the dataset are numeric, but there's a few categorical variables I'll be exploring as well.  

Worst &rarr; Best <br>
ProsperCreditGrade: NC, HR, E, D, C, B, A, AA, N/A <br>
* NC means No Credit and is only present in loans originated before July 2009. N/A means there wasn't a credit grade assigned; potentially because of previous accounts with Prosper Loans. HR means High Risk and E &rarr; AA are credit grades in ascending order.

LoanStatus: Canceled, Chargedoff, Defaulted, Past Due (>120 days), Past Due (91-120 days), Past Due (61-90 days), Past Due (31-60 days), Past Due (16-30 days), Past Due (1-15 days), Current, FinalPaymentInProgress, Completed <br> <br>
Shortest &rarr; Longest <br>
Term: 1 Year, 3 Years, 5 Years

### What is/are the main feature(s) of interest in your dataset?

> I'd like to see how loan amount & interest rate differ between different types of borrowers.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> I think the most useful features to explore will be credit grades, credit scores, occupation, income, 

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!